src/polly/src/epoll.rs

```diff
@@ -0,0 +1,201 @@
+    }
+}
+
+/// This is a wrapper over 'libc::epoll_event'.
+///
+/// We are using `transparent` here to be super sure that this struct and its fields
+/// have the same alignment as those from the `epoll_event` struct from C.
+#[repr(transparent)]
+#[derive(Clone, Copy)]
+pub struct Event(epoll_event);
+
+impl Deref for Event {
+    type Target = epoll_event;
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl Default for Event {
+    fn default() -> Self {
+        Event(epoll_event {
+            events: 0u32,
+            u64: 0u64,
+        })
+    }
+}
+
+impl Event {
+    /// Create a new epoll_event instance with the following fields: `events`, which contains
+    /// an event mask and `data` which represents a user data variable. `data` field can be
+    /// a fd on which we want to monitor the events specified by `events`.
+    pub fn new(events: EventType, data: u64) -> Self {
+        Event(epoll_event {
+            events: events.bits(),
+            u64: data,
+        })
+    }
+
+    pub fn events(&self) -> u32 {
+        self.events
+    }
+
+    pub fn data(&self) -> u64 {
+        self.u64
+    }
+}
+
+/// Stores a file descriptor referring to an epoll instance.
+#[derive(Clone, Copy, Debug)]
+pub struct Epoll {
+    epoll_fd: RawFd,
+}
+
+impl Epoll {
+    /// Create a new epoll file descriptor.
+    pub fn new() -> io::Result<Self> {
+        let epoll_fd = SyscallReturnCode(unsafe { epoll_create1(EPOLL_CLOEXEC) }).into_result()?;
+        Ok(Epoll { epoll_fd })
+    }
+
+    /// Wrapper for `libc::epoll_ctl`.
+    ///
+    /// This can be used for adding, modifying or removing a file descriptor in the
+    /// interest list of the epoll instance.
+    ///
+    /// # Arguments
+    ///
+    /// * `operation` refers to the action to be performed on the file descriptor.
+    /// * `fd` is the file descriptor on which we want to perform `operation`.
+    /// * `event` refers to the `epoll_event` instance that is linked to `fd`.
+    pub fn ctl(self, operation: ControlOperation, fd: RawFd, mut event: Event) -> io::Result<()> {
+        // Safe because we give a valid epoll file descriptor, a valid file descriptor to watch,
+        // as well as a valid epoll_event structure. We also check the return value.
+        SyscallReturnCode(unsafe {
+            epoll_ctl(
+                self.epoll_fd,
+                operation as i32,
+                fd,
+                &mut event as *mut _ as *mut epoll_event,
+            )
+        })
+        .into_empty_result()
+    }
+
+    /// Wrapper for `libc::epoll_wait`.
+    ///
+    /// # Arguments
+    ///
+    /// * `events` points to a memory area that will be used for storing the events
+    /// returned by `epoll_wait()` call.
+    /// * `max_events` is the maximum number of events that we want to be stored in
+    /// `events` buffer.
+    /// * `timeout` specifies for how long the `epoll_wait` system call will block
```

Comment:
Also add a comment about what is returned. You can add it at the beginning of the doc comment after: "Wrapper for .."